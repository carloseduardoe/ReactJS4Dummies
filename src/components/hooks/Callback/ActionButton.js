import React from 'react';

const ActionButton = ({text, action}) => {
    console.log("button", text);
    return <button onClick={action}>{text}</button>;
}

export default React.memo(ActionButton);
