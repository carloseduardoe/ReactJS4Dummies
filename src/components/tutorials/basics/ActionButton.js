import React, { Component } from 'react'

class ActionButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.initialValue
    };
  }

  performAction() {
    this.setState((prevState) => ({ value: !prevState.value }));
  }

  render() {
    return <div>
      <h6 data-testid="action-button_message">{`Value is ${this.state.value}`}</h6>
      <br/>
      <button data-testid="action-button_button" onClick={() => this.performAction()}>Toggle</button>
    </div>;
  }
}

export default ActionButton