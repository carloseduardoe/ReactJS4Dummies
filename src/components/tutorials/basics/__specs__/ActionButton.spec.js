import React from 'react';
import ReactDOM from 'react-dom';
import ActionButton from '../ActionButton';

import { render, cleanup } from '@testing-library/react';
import { create, act } from "react-test-renderer";
import '@testing-library/jest-dom';

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
  cleanup();
});

describe("Action Button", () => {
    const value = false;

    it('renders', () => {
        ReactDOM.render(
            <ActionButton initialValue={value}/>,
            container
        );
    });

    it('performs action', () => {
        let component, button, message;
        act(() => {
            component = create(<ActionButton initialValue={value}/>);
        });

        // Using react-test-renderer
        const testInstance = component.root;
        const componentInstance = component.getInstance();
        message = testInstance.findByType("h6");
        button = testInstance.findByType("button");
        
        expect(componentInstance.state.value).toBe(value);
        act(() => button.props.onClick());
        expect(componentInstance.state.value).toBe(!value);
        act(() => button.props.onClick());
        expect(componentInstance.state.value).toBe(value);
        expect(message.props.children).toBe(`Value is ${value}`);


        // Using @testing-library
        const { getByTestId } = render(<ActionButton initialValue={value}/>);
        message = getByTestId("action-button_message");
        button = getByTestId("action-button_button");

        expect(message.textContent).toBe(`Value is ${value}`);
        act(() => {button.click()});
        expect(message.textContent).toBe(`Value is ${!value}`);
        act(() => {button.click()});
        expect(message.textContent).toBe(`Value is ${value}`);


        // Using Web APIs
        ReactDOM.render(<ActionButton initialValue={value}/>, container);
        message = container.getElementsByTagName("h6")[0];
        button = container.getElementsByTagName("button")[0];

        expect(message.textContent).toBe(`Value is ${value}`);
        act(() => button.click());
        expect(message.textContent).toBe(`Value is ${!value}`);
        act(() => button.click());
        expect(message.textContent).toBe(`Value is ${value}`);
    });

    it('matches snapshot', () => {
        const tree = create(<ActionButton initialValue={value}/>).toJSON();

        expect(tree).toMatchSnapshot();
    });
});