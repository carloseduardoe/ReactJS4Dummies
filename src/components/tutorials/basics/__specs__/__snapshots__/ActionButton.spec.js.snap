// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Button matches snapshot 1`] = `
<div>
  <h6
    data-testid="action-button_message"
  >
    Value is false
  </h6>
  <br />
  <button
    data-testid="action-button_button"
    onClick={[Function]}
  >
    Toggle
  </button>
</div>
`;
